# Building Handwritten Mathematical Expressions Recognition Model
### Khiem T. Do, Phuong T.M. Chu & Santhos

In this notebook, we demonstrate how to build our Handwritten Mathematical Expressions Recognition models for different Operators.

In [1]:
from __future__ import absolute_import, division, print_function, unicode_literals

import matplotlib.pylab as plt
import shutil # copy, move file
import os # miscellaneous operation system interfaces
import pathlib
import random
import numpy as np
import pandas as pd
import seaborn as sns
from PIL import Image
%matplotlib inline

/home/reichu2/anaconda3/lib/python3.7/site-packages/sklearn/externals/joblib/__init__.py:15: DeprecationWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=DeprecationWarning)


In [53]:
data_root = pathlib.Path('train')

In [54]:
# Find all image's path

all_image_paths = list(data_root.glob("*/*"))
all_image_paths = [str(path) for path in all_image_paths if path.is_file()]
random.shuffle(all_image_paths)

image_count = len(all_image_paths)
image_count

12119

In [ ]:
for index, image_path in enumerate(all_image_paths):
    try:
        image = tf.io.read_file(image_path)
        image = tf.image.decode_jpeg(image, channels=3)
    except:
        print(all_image_paths[index])
    if index % 200 == 0: 
        print(index)

In [56]:
# Find all label names

label_names = sorted(item.name for item in data_root.glob('*/') if item.is_dir())
label_names = np.array(label_names)
label_names

array(['*', '+', '-', 'div'], dtype='<U3')

In [57]:
label_to_index = dict((name, index) for index, name in enumerate(label_names))

label_to_index

{'*': 0, '+': 1, '-': 2, 'div': 3}

In [58]:
# Find all image's labels

all_image_labels = [label_to_index[pathlib.Path(path).parent.name]
                    for path in all_image_paths]

print("First 10 labels indices: ", all_image_labels[:10])

First 10 labels indices:  [2, 1, 0, 1, 2, 2, 1, 1, 1, 2]


In [59]:
# Count #image each class

from collections import Counter
count = Counter()

for label in all_image_labels:
    count[label_names[label]] += 1
  
count

Counter({'-': 4000, '+': 4000, '*': 3251, 'div': 868})

In [60]:
df = pd.DataFrame(columns=['path','label'])
df['path'] = all_image_paths
df['label'] = all_image_labels

In [61]:
mnist = pd.read_csv("train.csv")

In [62]:
newdf = pd.DataFrame(columns=mnist.columns)

In [63]:
newdf.drop('label',axis=1,inplace=True)

In [64]:
newdf.head()

Empty DataFrame
Columns: [pixel0, pixel1, pixel2, pixel3, pixel4, pixel5, pixel6, pixel7, pixel8, pixel9, pixel10, pixel11, pixel12, pixel13, pixel14, pixel15, pixel16, pixel17, pixel18, pixel19, pixel20, pixel21, pixel22, pixel23, pixel24, pixel25, pixel26, pixel27, pixel28, pixel29, pixel30, pixel31, pixel32, pixel33, pixel34, pixel35, pixel36, pixel37, pixel38, pixel39, pixel40, pixel41, pixel42, pixel43, pixel44, pixel45, pixel46, pixel47, pixel48, pixel49, pixel50, pixel51, pixel52, pixel53, pixel54, pixel55, pixel56, pixel57, pixel58, pixel59, pixel60, pixel61, pixel62, pixel63, pixel64, pixel65, pixel66, pixel67, pixel68, pixel69, pixel70, pixel71, pixel72, pixel73, pixel74, pixel75, pixel76, pixel77, pixel78, pixel79, pixel80, pixel81, pixel82, pixel83, pixel84, pixel85, pixel86, pixel87, pixel88, pixel89, pixel90, pixel91, pixel92, pixel93, pixel94, pixel95, pixel96, pixel97, pixel98, pixel99, ...]
Index: []

[0 rows x 784 columns]

In [ ]:
from PIL import Image
for i in range(df.shape[0]):
    img = Image.open(df.loc[i,'path'])
    img.load()
    img = img.resize((28, 28), Image.BILINEAR)
    img = np.asarray( img, dtype="int64" )
    img = img.ravel()
#   Chang to black background
    img = 255 - img
    newdf.loc[i,:]=img
    if i % 200 == 0: 
        print(i)

In [67]:
label = pd.DataFrame(columns=['label'])
label['label'] = df['label']
data = label.join(newdf)

In [68]:
#   Save a flattern dataframe
data.to_csv('data.csv')

## Building Model

In [3]:
# Import the modules
from sklearn.externals import joblib
from sklearn import datasets
from skimage.feature import hog
from sklearn.svm import LinearSVC
import numpy as np
from collections import Counter
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC

In [4]:
df = pd.read_csv(r"data.csv")

In [5]:
df = df.drop('Unnamed: 0', axis=1)
df.head()

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      2     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1      1     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2      0     1.0     1.0     1.0     1.0     1.0     1.0     1.0     0.0   
3      1     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4      2     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0     0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
1     0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
2     1.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
3     0.0  ...       1.0       1.0       0.0       0.0       0.0       0.0   
4     0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   

   pixel780  pixel781  pixel782  pixel783  
0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0  

[5 rows x 785 columns]

In [6]:
# Extract the features and labels
X = df.drop('label', axis=1).to_numpy()
y = df['label'].to_numpy()

In [7]:
df['label'].value_counts()

2    4000
1    4000
0    3251
3     868
Name: label, dtype: int64

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.1, random_state = 101)

In [11]:
# Extract the hog features

def ExtractHog(features):
    list_hog_fd = []
    for feature in features:
        fd = hog(feature.reshape((28, 28)), orientations=9, pixels_per_cell=(14, 14), cells_per_block=(1, 1), visualize=False)
        list_hog_fd.append(fd)
    hog_features = np.array(list_hog_fd, 'float64')
    return hog_features

In [12]:
X_train_hog = ExtractHog(X_train)
X_test_hog = ExtractHog(X_test)

### Tunning hyperparameters

In [ ]:
from sklearn.model_selection import cross_val_score

val_rate = []
c_range =  range(1,200,20)

for i in c_range:
    
    svm = SVC(C=i, kernel='linear')
    val_error = 1 - cross_val_score(svm, X_train_hog, y_train,cv=5).mean()
    val_rate.append(val_error)


# Plot settings
plt.figure(figsize=(15,7))
plt.plot(c_range, val_rate, color='orange', linestyle='dashed', marker='o',
         markerfacecolor='black', markersize=5, label='Validation Error')

plt.xticks(np.arange(c_range.start, c_range.stop, c_range.step), rotation=60)
plt.grid()
plt.legend()
plt.title('Validation Error vs. C Value')
plt.xlabel('C')
plt.ylabel('Validation Error')
plt.show()

In [ ]:
best_c = c_range[val_rate.index(min(val_rate))]
best_c

### Building Linear SVM model

In [14]:
# Create an linear SVM object
clf = LinearSVC()

In [15]:
# Perform the training
clf.fit(X_train_hog, y_train)

LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
          intercept_scaling=1, loss='squared_hinge', max_iter=1000,
          multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
          verbose=0)

### Evaluate the accuracy of model

In [16]:
clf_predictions = clf.predict(X_test_hog)

clf_acc = accuracy_score(clf_predictions, y_test)
print(clf_acc)

0.9488448844884488


In [17]:
# Save the classifier
joblib.dump(clf, "model_cls_operator.pkl", compress=3)

['model_cls_operator.pkl']

In [18]:
model_operator = joblib.load("model_cls_operator.pkl")

labels_name = ['*', '+', '-', 'div']

In [19]:
clf_predictions = model_operator.predict(X_test_hog)

clf_acc = accuracy_score(clf_predictions, y_test)
print(clf_acc)

0.9488448844884488


In [22]:
clf_predictions.max()

3

In [23]:
clf_predictions.min()

0

## Summary
We successfully built a SVC model to predict handwritten mathematical operators with the accuracy of 94.88% 